src/event-stream/tokens-contract-handler.ts

```diff
@@ -626,27 +648,76 @@ class TokensContractHandler {
   /**
    * Make readonly contract call
    */
-  private async makeReadOnlyContractCall(
+  async makeReadOnlyContractCall(
+    functionName: string,
+    functionArgs: ClarityValue[]
+  ): Promise<ReadOnlyContractCallResponse> {
+    const url = this.stacksNetwork.getReadOnlyFunctionCallApiUrl(
+      this.contractAddress,
+      this.contractName,
+      functionName
+    );
+    const args = functionArgs.map(arg => cvToHex(arg));
+
+    const body = JSON.stringify({
+      sender: this.address,
+      arguments: args,
+    });
+
+    let response: Response;
+
+    try {
+      response = await fetchPrivate(url, {
+        method: 'POST',
+        body,
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+    } catch {
+      // In case there is a network error during fetch.
+      return { found: false, retriable: true };
+    }
+
+    if (!response.ok) {
+      if (response.status >= 500) return { found: false, retriable: true };
```

Comment:
The lack of comments around this condition requires reviewers to do a lot of research. It's very unclear to me if this check is correct, too aggressive, or not aggressive enough.

 For example:
* What happens if the stacks-node RPC server is responsive, but hasn't yet synced to the block with this smart-contract? Is there a 2xx OK status but error in the body? Is there a non-2xx status but indication that the error is temporary until synced? Etc.
* Can the stacks-node ever respond with 2xx OK status but with an http response payload indicating some kind of error condition?
* What is the http response for a readonly contract-call request that was successfully processed by the stacks-node RPC server and successfully evaluated by the Clarity VM, but resulted in a Clarity function error result? Does it return 2xx OK but have an error response body?


At the least, these need tested and documented. Ideally there would also be integration tests against the stacks-node as well.

Reply:
Yes, as long as the server is responding, it will return 200. I have also added a test without deploying the contract call. It throws an error with the message `Unchecked(NoSuchContract({contract_id))` [here](https://github.com/hirosystems/stacks-blockchain-api/pull/1131/files#diff-1be0acee815b65d2c2026b016d6bfc8eff2e08f184b79c128001d50680f71421R184-R203).

Yes, the same test is applicable for 2nd question. It returns headers with 200 and OK. However, fails during `response.json()` since the contract doesn't exist.

Added another [test](https://github.com/hirosystems/stacks-blockchain-api/pull/1131/files#diff-1be0acee815b65d2c2026b016d6bfc8eff2e08f184b79c128001d50680f71421R156-R180) for testing the third question. The contract function returns the object with `ClarityType.ResponseErr` as type and value for it.